woocommerce/woocommerce
# WooCommerce Docs: Store post meta of edit_url, refactor ManifestProcessor for testability

### Changes

We want to support docs having an edit url link. This link will be used to direct users to Github where they can edit a document at its original location much like other technical documentation websites.

To do this I decided to use post meta which makes it the most flexible for the implementers of any theme or front-end for WooCommerce docs, so they can display the edit url as they see fit rather than putting it directly in the post content.

Also as part of this (as I discussed with @lsinger in a previous PR) I refactored out some of the functionality of ManifestProcessor to separate classes. ManifestProcessor now mostly just coordinates traversing the manifest, downloading files and it does call the WP set categories method, but thats all. This allowed me to cover the behaviours with more unit tests directly instead of relying on a single integration test of the ManifestProcessor.

### Testing

In an ideal world the new testing should be enough, if it passes I'm pretty confident at this stage that the system is working. But in reality still smoke testing the manifest behaviour is worthwhile.

1. Have the docs plugin mounted in your dev environment.
2. Build the docs plugin, via pnpm start
3. Now visit the menu item for the plugin in wp-admin (called WooCommerce Docs)
4. Add a new manifest URL via the basic UI, it will need to point to the sample manifest that is included with the plugin. 
6. The URL should be like http://your-dev-hostname/wp-content/plugins/woocommerce-docs/scripts/manifest.json
The action scheduler job runs once a minute after activation of the plugin. Once it has run it should have generated a set of categories and posts with content matching that of the sample docs in example-docs folder that is bundled with the plugin currently.
7. To see the post meta you can query it in your SQL tool of choice (let me know if you need help setting this up): `SELECT * from wp_postmeta;` should show you the post meta thats been generated.
